internal/search/crawler/github/crawler.go

```diff
@@ -0,0 +1,420 @@
+// Implements crawler.Crawler.
+func (gc githubCrawler) Crawl(
+	ctx context.Context, output chan<- *doc.KustomizationDocument) error {
+
+	// Range finding will be added in the next PR to make this one
+	// simpler/shorter. It would return multiple search queries for the
+	// Github API such that all documents can be retrieved. This is
+	// required since Github returns a max of 1000 results per query, so
+	// multiple queries that split the search space into chunks of 1000
+	// kustomization files is required.
+	ranges := []string{gc.query.String()}
+
+	// Query each range for files.
+	errs := make(multiError, 0)
+	for _, query := range ranges {
+		err := ProcessQuery(ctx, gc.rc, query, output)
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	return errs
+}
+
+// ProcessQuery follows all of the pages in a query, and updates/adds the
+// documents from the crawl to the datastore/index.
+func ProcessQuery(ctx context.Context, rc RequestConfig, query string,
+	output chan<- *doc.KustomizationDocument) error {
+
+	queryPages := make(chan GithubResponseInfo)
+
+	go func() {
+		// Forward the document metadata to the retrieval channel.
+		// This separation allows for concurrent requests for the code
+		// search, and the retrieval portions of the API.
+		err := ForwardPaginatedQuery(
+			ctx, query, rc.retryCount, queryPages)
+		if err != nil {
+			// TODO(damienr74) handle this error with redis?
+			logger.Println(err)
+		}
+		close(queryPages)
+	}()
+
+	errs := make(multiError, 0)
+	for page := range queryPages {
+		if page.Error != nil {
+			errs = append(errs, page.Error)
+			continue
+		}
+
+		for _, file := range page.Parsed.Items {
+			// TODO(damienr74) This is where we'd neet to
+			// communicate with redis. Currently always doing a full
+			// reindex of the documents. Since the documents are in
+			// sorted order in each bucket, we can short circuit the
+			// search when we find a file that has been seen, or we
+			// can choose to selectively update files.
+
+			k, err := kustomizationResultAdapter(rc, file)
+			if err != nil {
+				errs = append(errs, err)
+				continue
+			}
+			output <- k
+		}
+	}
+
+	return errs
+}
+
+func kustomizationResultAdapter(rc RequestConfig, k GithubFileSpec) (
+	*doc.KustomizationDocument, error) {
+
+	data, err := GetFileData(rc, k)
+	if err != nil {
+		return nil, err
+	}
+
+	creationTime, err := GetFileCreationTime(rc, k)
+	if err != nil {
+		logger.Printf("(Error: %v) initializing to current time.", err)
+	}
+
+	doc := doc.KustomizationDocument{
+		DocumentData:  string(data),
+		FilePath:      doc.Atom(k.Path),
+		RepositoryURL: doc.Atom(k.Repository.URL),
+		CreationTime:  creationTime,
+	}
+
+	return &doc, nil
+}
+
+// ForwardPaginatedQuery follows the links to the next pages and performs all of
+// the queries for a given search query, relaying the data from each request
+// back to an output channel.
+func ForwardPaginatedQuery(ctx context.Context, query string, retryCount uint64,
```

Comment:
I'm curious have you considered using some existing library for github API, such as https://github.com/google/go-github. What's the reason of not using that.

Reply:
It just seems like a huge library to import for the tiny bit of it that I'm using, but if you'd prefer that I use the library I could. It should be pretty easy to change!

Reply:
Not necessary. I wanted to understand why not using it.